src/sinks/gcp/stackdriver/logs/tests.rs

```diff
@@ -0,0 +1,314 @@
+            }
+        })
+    );
+}
+
+#[test]
+fn encode_inserts_timestamp() {
+    let transformer = Transformer::default();
+
+    let encoder = StackdriverLogsEncoder::new(
+        transformer,
+        Template::try_from("testlogs").unwrap(),
+        StackdriverLogName::Project("project".to_owned()),
+        StackdriverResource {
+            type_: "generic_node".to_owned(),
+            labels: HashMap::from([(
+                "namespace".to_owned(),
+                Template::try_from("office").unwrap(),
+            )]),
+        },
+        Some(ConfigValuePath::try_from("anumber".to_owned()).unwrap()),
+    );
+
+    let mut log = LogEvent::default();
+    log.insert("message", Value::Bytes("hello world".into()));
+    log.insert("anumber", Value::Bytes("100".into()));
+    log.insert(
+        "timestamp",
+        Value::Timestamp(
+            Utc.with_ymd_and_hms(2020, 1, 1, 12, 30, 0)
+                .single()
+                .expect("invalid timestamp"),
+        ),
+    );
+
+    let json = encoder.encode_event(Event::from(log)).unwrap();
+    assert_eq!(
+        json,
+        serde_json::json!({
+            "logName":"projects/project/logs/testlogs",
+            "jsonPayload":{"message":"hello world","timestamp":"2020-01-01T12:30:00Z"},
+            "severity":100,
+            "resource":{
+                "type":"generic_node",
+                "labels":{"namespace":"office"}},
+            "timestamp":"2020-01-01T12:30:00Z"
+        })
+    );
+}
+
+#[test]
+fn severity_remaps_strings() {
+    for &(s, n) in &[
+        ("EMERGENCY", 800), // Handles full upper case
+        ("EMERG", 800),     // Handles abbreviations
+        ("FATAL", 800),     // Handles highest alternate
+        ("alert", 700),     // Handles lower case
+        ("CrIt1c", 600),    // Handles mixed case and suffixes
+        ("err404", 500),    // Handles lower case and suffixes
+        ("warnings", 400),
+        ("notice", 300),
+        ("info", 200),
+        ("DEBUG2", 100), // Handles upper case and suffixes
+        ("trace", 100),  // Handles lowest alternate
+        ("nothing", 0),  // Maps unknown terms to DEFAULT
+        ("123", 100),    // Handles numbers in strings
+        ("-100", 0),     // Maps negatives to DEFAULT
+    ] {
+        assert_eq!(
+            remap_severity(s.into()),
+            Value::Integer(n),
+            "remap_severity({:?}) != {}",
+            s,
+            n
+        );
+    }
+}
+
+async fn build_request(
+    auth: GcpAuthenticator,
+    uri: Uri,
+    events: Vec<BoxedRawValue>,
+) -> crate::Result<Request<Bytes>> {
+    let events = serde_json::json!({ "entries": events });
+
+    let body = crate::serde::json::to_bytes(&events).unwrap().freeze();
+
+    let mut request = Request::post(uri.clone())
+        .header("Content-Type", "application/json")
+        .body(body)
+        .unwrap();
+    auth.apply(&mut request);
+
+    Ok(request)
+}
```

Comment:
The `correct_request` test is now primarily testing the logic of this test helper function, rather than the actual implementation.  Could we refactor this test so that it tests the real implementation? We should be able to do so with a combination of the `StackdriverLogsEncoder` and `StackdriverLogsSinkRequestBuilder`.